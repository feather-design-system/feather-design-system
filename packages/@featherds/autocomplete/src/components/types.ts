import { Ref, PropType } from "vue";
import HighlightProps from "./Highlight/HighlightProps";
import { InputWrapperProps, InputSubTextProps } from "@featherds/input-helper";
interface IAutocompleteItem extends IAutocompleteItemType {
  _new?: string;
  _text: string;
  _pre?: IAutocompleteChipIcon;
}
interface IAutocompleteItemType {
  [k: string]: unknown;
}

interface IAutocompleteGridColumn {
  prop: keyof IAutocompleteItemType;
  title: string;
  align?: "right" | "left" | "center";
}

interface IAutocompleteResultRender {
  reset: () => void;
  handleKeyPress: (
    e: KeyboardEvent,
    internalResults: IAutocompleteItemType[]
  ) => boolean;
  active: {
    row: number;
    col?: number;
  };
  first: () => void;
}

export enum AutocompleteTypes {
  multi = "multi",
  single = "single",
}
interface IStrategyOptions {
  selectionLimit: Ref<number>;
  selectionLimitReached: Ref<boolean>;
  modelValue: Ref<IAutocompleteItemType | IAutocompleteItemType[]>;
  textProp: Ref<string>;
  input: Ref<HTMLInputElement>;
  query: Ref<string>;
  emitSearch: () => void;
  forceCloseResults: Ref<boolean>;
  allowNew: Ref<boolean>;
  internalResults: Ref<IAutocompleteItemType[]>;
}

interface IAutocompleteChipIcon {
  icon: unknown;
  title: string;
}

export type {
  IStrategyOptions,
  IAutocompleteGridColumn,
  IAutocompleteItemType,
  IAutocompleteItem,
  IAutocompleteChipIcon,
  IAutocompleteResultRender,
};

export const LABELS = {
  noResults: "No results found",
  minChar: "Enter ${min} characters to search",
  clear: "Clear value",
  selectionLimit: "Selection limit reached",
  new: "New",
  remove: "Remove",
};
export const props = {
  id: {
    type: String,
  },
  modelValue: {
    type: [Array, Object] as PropType<
      IAutocompleteItemType[] | IAutocompleteItemType
    >,
  },
  results: {
    type: Array as PropType<IAutocompleteItemType[]>,
    default: () => [],
  },
  textProp: {
    type: String as unknown as PropType<keyof IAutocompleteItemType>,
    default: "_text",
  },
  loading: {
    type: Boolean,
    default: false,
  },
  minChar: {
    type: Number,
    default: 0,
  },
  allowNew: {
    type: Boolean,
    default: false,
  },
  selectionLimit: {
    type: Number,
  },
  newMatcher: {
    type: Function,
    default: (
      item: IAutocompleteItemType,
      query: string,
      comp: { textProp: string }
    ) => {
      return (
        (item[comp.textProp] as string).toString().toLowerCase() ===
        query.toLowerCase()
      );
    },
  },
  type: {
    type: String as PropType<keyof typeof AutocompleteTypes>,
    required: true,
    validator: (v: keyof typeof AutocompleteTypes) => {
      // The value must match either
      return !!AutocompleteTypes[v];
    },
  },
  labels: {
    type: Object as PropType<Partial<typeof LABELS>>,
    default() {
      return LABELS;
    },
  },
  gridConfig: {
    type: Array as PropType<IAutocompleteGridColumn[]>,
  },
  schema: {
    type: Object,
    required: false,
  },
  ...HighlightProps,
  ...InputSubTextProps,
  ...InputWrapperProps,
} as const;

export const emits = {
  "update:modelValue": (
    value: IAutocompleteItemType | IAutocompleteItemType[] | undefined
  ) => true,
  new: (value: string) => true,
  search: (value: string) => true,
};
