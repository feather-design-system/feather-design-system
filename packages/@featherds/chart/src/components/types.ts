export type ShirtSize = "xs" | "sm" | "md" | "lg" | "xl";

export type FeatherChartType = "bar";

export type FeatherBarChartData = {
  labels: Array<string>;
  datasets: Array<{
    label: string;
    data: Array<string | number>;
    backgroundColor?: string;
  }>;
};

export type FeatherBarChartOptions = {
  responsive?: boolean;
  maintainAspectRation?: boolean;
};

export type FeatherD3Dimensions = {
  height: number;
  width: number;
  margin: {
    top: number;
    right: number;
    bottom: number;
    left: number;
  }
};

export type FeatherD3Options = {
  units: string;
  class?: "color1" | "color2";
};


// Data can be any array of objects returned from an API or local JSON
export type FeatherD3Data = Array<unknown>;

export interface ICoordinateAxes {
  x: string,
  y: string
}
