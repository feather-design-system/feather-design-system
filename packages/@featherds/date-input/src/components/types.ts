export interface IDateDisabledConfig extends IDateRange {
  dates?: (Date | number)[];
  ranges?: IDateRange[];
  days?: number[];
  daysOfMonth?: number[];
  customPredictor?: (date: Date | number) => boolean;
}
interface IDateRange {
  to?: Date | number;
  from?: Date | number;
}
export const LABELS = {
  day: "Day",
  month: "Month",
  year: "Year",
  prevMonth: "Previous month",
  nextMonth: "Next month",
  selectMonth: "Select month",
  selectYear: "Select year",
  inCalendar: "In calendar",
  calendar: "Click for calendar",
  clear: "Clear date",
  menu: "Use arrow keys to navigate dates. Page down and page up will navigate by month. Shift page down and shift page up will navigate by year. Press escape to exit the calendar.",
};
