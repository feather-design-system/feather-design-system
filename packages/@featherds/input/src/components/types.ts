import { InputWrapperProps, InputSubTextProps } from "@featherds/input-helper";
import { PropType } from "vue";
export const props = {
  type: {
    type: String,
    default: "text",
  },
  modelValue: {
    type: [String, Number] as PropType<string | number>,
  },
  maxlength: {
    type: Number,
    required: false,
    default: 0,
  },
  schema: {
    type: Object as PropType<Record<string, any>>,
    required: false,
  },
  id: {
    type: String,
    required: false,
  },
  ...InputWrapperProps,
  ...InputSubTextProps,
} as const;

export const emits = {
  "update:modelValue": (value: string | undefined | number) => true,
};
