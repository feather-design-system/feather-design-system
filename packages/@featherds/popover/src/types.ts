// Note: introduced new types
export type Tooltip = {
  height: number;
  width: number;
};

export type Trigger = {
  height: number;
  width: number;
  left: number;
  top: number;
};

export enum PopoverPlacement {
  top = "top",
  bottom = "bottom",
  left = "left",
  right = "right",
}

export enum PointerAlignment {
  center = "center",
  left = "left",
  right = "right",
}
