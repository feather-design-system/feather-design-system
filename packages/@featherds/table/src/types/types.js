export const ALIGNMENT = {
  LEFT: "left",
  RIGHT: "right",
  CENTER: "center",
};

export const SORT = {
  ASCENDING: "asc",
  DESCENDING: "desc",
  NONE: "none",
};
