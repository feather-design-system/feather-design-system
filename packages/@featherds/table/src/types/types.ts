export enum ALIGNMENT {
  LEFT = "left",
  RIGHT = "right",
  CENTER = "center",
}

export enum SORT {
  ASCENDING = "asc",
  DESCENDING = "desc",
  NONE = "none",
}
