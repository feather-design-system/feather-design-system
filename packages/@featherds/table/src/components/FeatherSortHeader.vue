<template>
  <th data-ref-id="feather-sort-header" :aria-sort="ariaSort">
    <div
      class="header-flex-container content"
      tabindex="0"
      @click="linkClicked"
      role="button"
      :aria-describedby="descriptionId"
    >
      <span class="sort-cell-label"><slot></slot></span>
      <span class="description">{{ sortDescriptionLabel }}</span>
      <span class="hidden-description" :id="descriptionId">{{
        clickSortLabel
      }}</span>

      <span class="icon focus hover" aria-hidden="true">
        <FeatherIcon :icon="sortIcon"></FeatherIcon>
        <FeatherRipple center />
      </span>
    </div>
    <span
      class="alert"
      role="alert"
      aria-live="assertive"
      aria-atomic="true"
      ref="alert"
    ></span>
  </th>
</template>
<script lang="ts">
import { FeatherIcon } from "@featherds/icon";
import { FeatherRipple } from "@featherds/ripple";
import ExpandLess from "@featherds/icon/navigation/ExpandLess";
import ExpandMore from "@featherds/icon/navigation/ExpandMore";
import UnfoldMore from "@featherds/icon/navigation/UnfoldMore";
import { SORT } from "../types/types";
import { getSafeId } from "@featherds/utils/id";
import { useLabelProperty } from "@featherds/composables/LabelProperty";
import { defineComponent, PropType, toRef } from "vue";
const LABELS = {
  sortAscending: "Sorted Ascending",
  sortDescending: "Sorted Descending",
  sortNone: "Sorted None",
  clickSort: "Click to Sort",
};
export const props = {
  sort: {
    type: String as PropType<SORT>,
    validator: function (value: SORT) {
      // The value must match either
      return (
        [SORT.ASCENDING, SORT.DESCENDING, SORT.NONE, "", undefined].indexOf(
          value
        ) !== -1
      );
    },
    required: true,
  },
  property: {
    type: String,
    required: true,
  },
  labels: {
    type: Object as PropType<Partial<typeof LABELS>>,
    default: () => {
      return LABELS;
    },
  },
} as const;
export const emits = {
  "sort-changed": (_v: { property: string; value: SORT }) => true,
};
export default defineComponent({
  emits: ["sort-changed"],
  props,
  setup(props) {
    return {
      ...useLabelProperty<typeof LABELS>(toRef(props, "labels"), LABELS),
    };
  },
  computed: {
    descriptionId() {
      return getSafeId("feather-sort-header-description");
    },
    sortIcon() {
      var result = UnfoldMore;
      if (this.sort === SORT.ASCENDING) {
        result = ExpandLess;
      }
      if (this.sort === SORT.DESCENDING) {
        result = ExpandMore;
      }
      return result;
    },
    ariaSort() {
      if (this.sort === SORT.ASCENDING) {
        return "ascending";
      } else if (this.sort === SORT.DESCENDING) {
        return "descending";
      } else {
        return "none";
      }
    },
    sortDescriptionLabel() {
      if (this.sort === SORT.NONE) {
        return this.sortNoneLabel;
      }
      return "";
    },
  },

  components: {
    FeatherIcon,
    FeatherRipple,
  },
  methods: {
    linkClicked: function (e: MouseEvent) {
      e.preventDefault();
      e.stopPropagation();
      // this is run when only the sort is to change
      var sortDir;
      switch (this.sort) {
        case SORT.ASCENDING:
          sortDir = SORT.DESCENDING;
          break;
        case SORT.DESCENDING:
          sortDir = SORT.NONE;
          break;
        default:
          sortDir = SORT.ASCENDING;
      }
      this.$emit("sort-changed", {
        property: this.property,
        value: sortDir,
      });
      this.announceSort(sortDir);
    },
    announceSort(dir: SORT) {
      let label = this.sortNoneLabel;

      if (dir === SORT.ASCENDING) {
        label = this.sortAscendingLabel;
      } else if (dir === SORT.DESCENDING) {
        label = this.sortDescendingLabel;
      }

      (this.$refs.alert as HTMLElement).textContent = label;
      setTimeout(() => {
        (this.$refs.alert as HTMLElement).textContent = "";
      }, 100);
    },
  },
});
</script>

<style lang="scss" scoped>
@import "@featherds/styles/themes/variables";
@import "@featherds/styles/themes/utils";
@import "@featherds/styles/mixins/typography";

th .header-flex-container {
  cursor: pointer;
  display: flex;
  align-items: center;
  font-weight: bold;
  &.content {
    @include state-on-surface();
  }
  .sort-cell-label {
    display: inline-block;
    user-select: none;
    -khtml-user-select: none;
    -moz-user-select: none;
    -ms-user-select: none;
    -o-user-select: none;
    -webkit-user-select: none;
  }

  &:focus {
    outline: 0;
  }
  span.icon {
    margin-left: 0.25rem;
    width: 1.5rem;
    height: 1.5rem;
    border-radius: 100%;
    position: relative;
    display: inline-flex;
    color: inherit;
    justify-content: center;
    align-items: center;
    flex: none;
    :deep(.feather-icon) {
      vertical-align: text-top;
      font-size: 1.125rem;
    }
    &:before,
    &:after {
      border-radius: 100%;
    }
  }
}
span.hidden-description {
  display: none;
}
span.description,
span.alert {
  @include screen-reader();
}
</style>
