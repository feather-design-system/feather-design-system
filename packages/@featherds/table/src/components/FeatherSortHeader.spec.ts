import { shallowMount } from "@vue/test-utils";
import FeatherSortHeader from "./FeatherSortHeader.vue";
import { SORT } from "../types/types";
import * as id from "@featherds/utils/id";
import { getCalls } from "@featherds/utils/test/calls";

jest.spyOn(id, "getSafeId").mockImplementation((x) => x);
const PROPERTY = "test";

const getWrapper = function (options: Record<string, unknown> = {}) {
  const props = (options.props as Record<string, unknown>) || {};
  options.props = {
    sort: SORT.ASCENDING,
    property: PROPERTY,
    ...props,
  };
  return shallowMount(FeatherSortHeader, options);
};

describe("FeatherSortHeader.vue", () => {
  it("should set aria-sort based on prop", async () => {
    const wrapper = getWrapper();
    expect(wrapper.element).toMatchSnapshot();
    await wrapper.setProps({
      sort: SORT.DESCENDING,
    });
    expect(wrapper.element).toMatchSnapshot();
    await wrapper.setProps({
      sort: SORT.NONE,
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should emit sort-changed when clicked", async () => {
    const wrapper = getWrapper();
    await wrapper.find("[role='button']").trigger("click");
    const firstEvent = getCalls<[{ property: string; value: SORT }]>(
      wrapper,
      "sort-changed"
    )[0][0];
    expect(firstEvent.property).toBe(PROPERTY);
    expect(firstEvent.value).toBe(SORT.DESCENDING);

    await wrapper.setProps({
      sort: SORT.DESCENDING,
    });
    await wrapper.find("[role='button']").trigger("click");
    const secondEvent = getCalls<[{ property: string; value: SORT }]>(
      wrapper,
      "sort-changed"
    )[1][0];
    expect(secondEvent.property).toBe(PROPERTY);
    expect(secondEvent.value).toBe(SORT.NONE);

    await wrapper.setProps({
      sort: SORT.NONE,
    });

    await wrapper.find("[role='button']").trigger("click");
    const thirdEvent = getCalls<[{ property: string; value: SORT }]>(
      wrapper,
      "sort-changed"
    )[2][0];
    expect(thirdEvent.property).toBe(PROPERTY);
    expect(thirdEvent.value).toBe(SORT.ASCENDING);
  });
});
