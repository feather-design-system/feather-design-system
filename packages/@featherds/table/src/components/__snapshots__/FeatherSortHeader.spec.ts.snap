// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeatherSortHeader.vue should set aria-sort based on prop 1`] = `
<th
  aria-sort="ascending"
  data-ref-id="feather-sort-header"
>
  <div
    aria-describedby="feather-sort-header-description"
    class="header-flex-container content"
    role="button"
    tabindex="0"
  >
    <span
      class="sort-cell-label"
    >
      
      
    </span>
    <span
      class="description"
    />
    <span
      class="hidden-description"
      id="feather-sort-header-description"
    >
      Click to Sort
    </span>
    <span
      aria-hidden="true"
      class="icon focus hover"
    >
      <feather-icon-stub
        flex="false"
        icon="[object Object]"
      />
      <feather-ripple-stub
        center="true"
      />
    </span>
  </div>
  <span
    aria-atomic="true"
    aria-live="assertive"
    class="alert"
    role="alert"
  />
</th>
`;

exports[`FeatherSortHeader.vue should set aria-sort based on prop 2`] = `
<th
  aria-sort="descending"
  data-ref-id="feather-sort-header"
>
  <div
    aria-describedby="feather-sort-header-description"
    class="header-flex-container content"
    role="button"
    tabindex="0"
  >
    <span
      class="sort-cell-label"
    >
      
      
    </span>
    <span
      class="description"
    />
    <span
      class="hidden-description"
      id="feather-sort-header-description"
    >
      Click to Sort
    </span>
    <span
      aria-hidden="true"
      class="icon focus hover"
    >
      <feather-icon-stub
        flex="false"
        icon="[object Object]"
      />
      <feather-ripple-stub
        center="true"
      />
    </span>
  </div>
  <span
    aria-atomic="true"
    aria-live="assertive"
    class="alert"
    role="alert"
  />
</th>
`;

exports[`FeatherSortHeader.vue should set aria-sort based on prop 3`] = `
<th
  aria-sort="none"
  data-ref-id="feather-sort-header"
>
  <div
    aria-describedby="feather-sort-header-description"
    class="header-flex-container content"
    role="button"
    tabindex="0"
  >
    <span
      class="sort-cell-label"
    >
      
      
    </span>
    <span
      class="description"
    >
      Sorted None
    </span>
    <span
      class="hidden-description"
      id="feather-sort-header-description"
    >
      Click to Sort
    </span>
    <span
      aria-hidden="true"
      class="icon focus hover"
    >
      <feather-icon-stub
        flex="false"
        icon="[object Object]"
      />
      <feather-ripple-stub
        center="true"
      />
    </span>
  </div>
  <span
    aria-atomic="true"
    aria-live="assertive"
    class="alert"
    role="alert"
  />
</th>
`;
